README.md
# sentry-traced

<p align="center"><img src="https://miro.medium.com/v2/resize:fit:4800/format:webp/1*CWU1u1K5-RtdDtZVobZxSA.png" width="500" height="auto"/></p>
<p align="center">
<a href="https://github.com/alexcambose/sentry-traced/actions/workflows/ci.yml">
<img src="https://github.com/alexcambose/sentry-traced/actions/workflows/ci.yml/badge.svg"/></a>
<a href="https://github.com/alexcambose/sentry-traced/issues"><img alt="GitHub issues" src="https://img.shields.io/github/issues/alexcambose/sentry-traced.svg?style=flat-square"></a>
<a href="https://www.npmjs.com/package/sentry-traced"><img alt="version" src="https://img.shields.io/npm/v/sentry-traced.svg?style=flat-square"></a>
<a href="https://github.com/alexcambose/sentry-traced/blob/master/LICENSE"><img alt="license" src="https://img.shields.io/github/license/alexcambose/sentry-traced.svg?style=flat-square"></a>
</p>

This package provides a decorator that can be used to annotate class methods with Sentry tracing information. The purpose of this decorator is to simplify the process of adding Sentry error reporting to a Node.js application, by automating the generation of Sentry tracing information for class methods.

See this article for more details about the implementation of this project: https://medium.com/@alexcambose/sentry-tracing-decorator-in-typescript-6d9607858ece

## Installation

To install the Sentry decorator package, simply run the following command in your terminal:

```
npm install sentry-traced
```

## Usage

To use the Sentry decorator, you must first import it into your project. You can do this by including the following line of code at the top of your file:

```ts
import { SentryTraced } from 'sentry-traced';
```

Next, you can use the `@SentryTraced` decorator to annotate class methods that you want to generate Sentry tracing information for.

```ts
class MyClass {
  @SentryTraced()
  myMethod() {
    // method implementation
  }
}
```

Once you have annotated your class methods with the @SentryTraced decorator, the Sentry tracing information will be automatically generated and sent to Sentry each time the annotated methods are called.

For example:
![Example1](./example/images/example1.png)

### Nested methods

As your codebase evolves, there will be some cases where a decorated function will be called by another decorated function. In this case one transaction will be generated (the root node) along with a tree like structure of spans (tree nodes) from the call stack.

For example, the following code will generate a structure that is interpreted by Sentry as this:

```ts
export class MainService {
  @SentryTraced()
  async myMethod() {
    await this.myMethodChild1();
    await this.myMethodChild2();
  }

  @SentryTraced()
  async myMethodChild1() {
    await this.myMethodChild1Child1();
    await this.myMethodChild2Child2();
  }

  @SentryTraced()
  async myMethodChild1Child1() {}

  @SentryTraced()
  async myMethodChild2Child2() {
    await this.myMethodChild2Child2Child1();
  }

  @SentryTraced()
  async myMethodChild2Child2Child1() {}

  @SentryTraced({ op: 'testOpParent' })
  async myMethodChild2() {
    const mainServiceChild = new MainServiceChild();
    mainServiceChild.mainServiceChildMethod();
  }
}

class MainServiceChild {
  @SentryTraced({ op: 'testOp', description: 'someDescription' })
  async mainServiceChildMethod() {}
}

const mainService = new MainService();
mainService.myMethod();
```

![Example2](./example/images/example2.png)

### Adding named parameters

The `sentry-traced` package provides the `@SentryParam()` decorator that can be used to annotate some parameters of a function.
This allows the values of these parameters to be added to the Sentry transaction as certain op.

```ts
class MainService {
  @SentryTraced()
  async myMethod(
    @SentryParam param1: number,
    @SentryParam param2: string,
    param3: string,
  ) {
    console.log('my method called', { param1, param2, param3 });
  }
}
```

![Example with params](./example/images/example3.png)

## Distributed tracing

Distributed tracing can be used to follow and analyze the path of requests as they move through various components or services. To achieven
this, the `sentry-traced` package provides the `withTracing` function that takes in the traceparent data and returns a function that can be used to wrap the code that needs to be traced.

Here's an example:

```ts
export class MainService {
  @SentryTraced()
  async myMethod(
    @SentryParam param1: number,
    @SentryParam param2: string,
    param3: string,
  ) {
    // get the current sentry span that surrounds this method
    const span = Sentry.getCurrentHub().getScope()?.getSpan();
    // call another endpoint and pass the traceparent header
    const result = await fetch('http://localhost:3000/other-endpoint', {
      headers: {
        'sentry-trace': span.toTraceparent(),
      },
    });
  }
}

export class SecondService {
  @SentryTraced()
  async myMethod(
    @SentryParam param1: number,
    @SentryParam param2: string,
    param3: string,
  ) {
    console.log('my second service method called', { param1, param2, param3 });
  }
}

const mainService = new MainService();
const secondService = new SecondService();

app.get('/', async (req, res) => {
  await mainService.myMethod(123, 'param2', 'param3');
  res.send('ok!');
});

app.get('/other-endpoint', async (req, res) => {
  // get the sentry traceparent header
  const sentryTrace = req.headers['sentry-trace'] as string;
  // wrap the second service method with `withTracing` and pass in the traceparent header
  // additionally you can pass in the operation name and description
  const result = withTracing(sentryTrace, {
    op: 'otherEndpoint',
    description: 'transaction from other endpoint',
  })(secondService.myMethod)(123, 'param2', 'param3');

  res.send('ok!');
});
```

the result transaction will appear in Sentry as follows:
![Example with params](./example/images/distributedTracing.png)

# Contributing

Contributions are welcome! Please check the contribution guidelines before submitting a pull request.

# Issues

If you encounter any issues or have suggestions, please open an issue.

# License

This project is licensed under the MIT License.

---

Feel free to reach out to me at alexcambose1@gmail.com if you have any questions or need further assistance.

Thanks for using **sentry-traced**!

:star:


CHANGELOG.md
## 1.1.2 (2023-05-29)

### Fix

- updated withTracing span metadata

## 1.1.1 (2023-05-27)

### Fix

- bug fixes

## 1.1.0 (2023-05-18)

### Feat

- added cz.toml
- updated example and readme
- general refactoring

### Fix

- linting issues


test/empty.test.ts
describe('empty test suite', () => {
  test('runs', () => {
    expect(true).toBe(true);
  });
});


test/index.test.ts
import { add, multiply } from '../';

describe('math functions', () => {
  test('adds two numbers', () => {
    expect(add(42, 3)).toBe(45);
  });

  test('multiplies three numbers', () => {
    expect(multiply(1, 2, 3)).toBe(6);
  });
});


src/test/utils.test.ts
import { generateSpanContext } from '../utils';

describe('utils', () => {
  describe('generateSpanContext', () => {
    it('correctly generates context for just method name', () => {
      expect(
        generateSpanContext({
          methodName: 'testMethod',
        }),
      ).toEqual({
        op: 'testMethod()',
        description: 'testMethod() call',
        descriptionNoArguments: 'testMethod() call',
        data: { args: undefined },
      });
    });
    it('correctly generates context for just class name', () => {
      expect(
        generateSpanContext({
          className: 'testClass',
        }),
      ).toEqual({
        op: 'testClass()',
        description: 'testClass() call',
        descriptionNoArguments: 'testClass() call',
        data: { args: undefined },
      });
    });
    it('correctly generates context for class name and method name without parameters', () => {
      expect(
        generateSpanContext({
          className: 'testClass',
          methodName: 'testMethod',
        }),
      ).toEqual({
        op: 'testClass.testMethod()',
        description: 'testClass.testMethod() call',
        descriptionNoArguments: 'testClass.testMethod() call',
        data: { args: undefined },
      });
    });
    it('correctly generates context for class name and method name with parameters', () => {
      expect(
        generateSpanContext({
          className: 'testClass',
          methodName: 'testMethod',
          args: [1, 'two', true],
        }),
      ).toEqual({
        op: 'testClass.testMethod(_,_,_)',
        description: 'testClass.testMethod(_,_,_) call',
        descriptionNoArguments: 'testClass.testMethod() call',
        data: { args: [1, 'two', true] },
      });
    });
    it('correctly generates context for class name and method name with parameters and Sentry traced params', () => {
      expect(
        generateSpanContext({
          className: 'testClass',
          methodName: 'testMethod',
          args: [1, 'two', true],
          sentryParams: [0, 2],
        }),
      ).toEqual({
        op: 'testClass.testMethod(1,_,true)',
        description: 'testClass.testMethod(1,_,true) call',
        descriptionNoArguments: 'testClass.testMethod() call',
        data: { args: [1, 'two', true] },
      });
    });
  });
});


src/utils.ts
import * as Sentry from '@sentry/node';
import { extractTraceparentData } from '@sentry/tracing';
import { SpanContext } from '@sentry/types';
import { InternalMetadata, SentryTracedParams } from './types';

declare global {
  // eslint-disable-next-line no-var
  var sentryTracedInstance: any;
}

export const registerSentryInstance = (sentryInstance: typeof Sentry) => {
  global.sentryTracedInstance = sentryInstance;
};

export const getSentryInstance = (): typeof Sentry => {
  return global.sentryTracedInstance;
};

/**
 * This function is used to check if a value is a promise
 * @param value Value to check if it's a promise
 * @returns Returns true if the value is a promise
 */
export const isPromise = (value: any): boolean => {
  return (
    value !== null &&
    typeof value === 'object' &&
    typeof value.then === 'function'
  );
};

/**
 * The data the Sentry needs to generate a span context
 * @param metadata Internal metadata used to generate the span context. It contains the class name, method name, arguments and the sentry params.
 * @param options Options related to the span context (eg. methodName, className)
 * @returns
 */
export const generateSpanContext = (
  metadata: InternalMetadata,
  options?: SentryTracedParams,
) => {
  const { className, methodName, args, sentryParams } = metadata;
  // this generates a string as a list of arguments, for example (1,2,3) or (1,_,3)
  let argumentsStringList = '()';
  if (args) {
    argumentsStringList = `(${new Array(args.length)
      .fill(0)
      .map((_, index) =>
        (sentryParams || []).includes(index) ? args[index] : '_',
      )
      .join(',')})`;
  }
  let functionNameString = `unknown`;
  if (!methodName && className) {
    functionNameString = className;
  } else if (!className && methodName) {
    functionNameString = methodName;
  } else {
    functionNameString = `${className}.${methodName}`;
  }

  const op = options?.op || `${functionNameString}${argumentsStringList}`;
  const description =
    options?.description || `${functionNameString}${argumentsStringList} call`;
  const descriptionNoArguments =
    options?.description || `${functionNameString}() call`;

  return {
    op,
    description,
    descriptionNoArguments,
    data: { args },
  };
};

/**
 * This function is used to connect a method annotated with `@SentryTraced` to an existing transaction based on the traceparentData string
 * See https://docs.sentry.io/platforms/node/performance/connect-services/
 * @param traceparentData The traceparentData string
 * @param overrides Options to pass to the transaction
 * @returns Returns a function that will wrap the original function that needs to be called
 * @example
 * ```typescript
 * const result = withTracing(sentryTrace)(secondService.myMethod)('some param1', 'some param2');
 * ```
 */
export const withTracing =
  (traceparentData?: string, overrides: SpanContext = {}) =>
  <T extends (...args: never[]) => any>(functionToCall: T) =>
  async (...args: Parameters<T>): Promise<Awaited<ReturnType<T>>> => {
    if (!traceparentData) {
      return await functionToCall.bind(functionToCall)(...args);
    }
    // The request headers sent by your upstream service to your backend.
    const extractedTraceparentData = extractTraceparentData(traceparentData);
    const className = functionToCall.constructor.name;
    const methodName = functionToCall.name;
    const { op, description } = generateSpanContext({
      className,
      methodName,
      args,
    });
    const transaction = Sentry.startTransaction({
      op,
      description,
      name: `${op} transaction`,
      ...extractedTraceparentData,
      ...overrides,
    });
    Sentry.configureScope((scope) => {
      scope.setSpan(transaction);
    });
    const result = await functionToCall.bind(functionToCall)(...args);
    transaction.finish();
    return result;
  };


src/SentryTraced.ts
import Sentry from '@sentry/node';
import 'reflect-metadata';
import { SentryTracedParams, SentryTracedParamsIndexes } from './types';
import { generateSpanContext, getSentryInstance, isPromise } from './utils';

const sentryParamsMetadataKey = Symbol('sentryParams');

/**
 * Decorator that automatically generates calls the sentry tracing related functions and registers nested aware metrics
 * @param options Decorator options related to sentry tracing namings
 * @returns Decorated function
 */
export const SentryTraced = (options?: SentryTracedParams) => {
  return (
    target: object,
    propertyKey: string,
    descriptor: PropertyDescriptor,
  ) => {
    const original = descriptor.value;
    descriptor.value = function (...args: unknown[]) {
      const className = this.constructor.name;
      const methodName = propertyKey;
      const sentryClient = getSentryInstance();
      const sentryParams: SentryTracedParamsIndexes =
        Reflect.getOwnMetadata(sentryParamsMetadataKey, target, methodName) ||
        [];
      const intermediaryFunction = async () => {
        if (!sentryClient || !sentryClient.getCurrentHub()) {
          throw new Error(`Sentry client not set`);
        }
        const spanContext = generateSpanContext(
          { className, methodName, args, sentryParams },
          options,
        );
        const scope = sentryClient.getCurrentHub().getScope();

        // get the current context, the order matters, we want to get
        // - the current span if it exists, this allows nesting if the function call is a child of a child or a leaf
        // - the current transaction if it exists, this can only happen if the function call is a child of the root node
        const contextTransaction = scope?.getSpan() || scope?.getTransaction();
        // we try to get the context span or transaction and if it doesn't exist we create a new transaction
        const transactionOrSpan =
          contextTransaction ||
          sentryClient.startTransaction({
            ...spanContext,
            name: spanContext.descriptionNoArguments,
          });
        // start a nested child of the current transaction/span
        const newSpan = transactionOrSpan?.startChild(spanContext);
        // set the context of the newly create span/transaction
        sentryClient.configureScope((scope) => {
          scope.setSpan(newSpan);
        });
        // call the function -> if the function is also annotated it will basically "recursively" create spans inside the parent span based on the context that we set
        const resultPromise = original.call(this, ...args);
        // there's a possibility that the function is a normal function or an async one so we need to check for both cases
        if (isPromise(resultPromise)) {
          // in case the function is a promise we need to resolve it and also reset the current context to the parent node for the next sibling function call
          return resultPromise.then((e: any) => {
            sentryClient.configureScope((scope) => {
              scope.setSpan(transactionOrSpan);
            });
            newSpan?.finish();
            if (!contextTransaction) {
              transactionOrSpan.finish();
            }
            return e;
          });
        } else {
          // same for a normal function but we just do it before returning the value
          Sentry.configureScope((scope) => {
            scope.setSpan(transactionOrSpan);
          });
          newSpan?.finish();
          if (!contextTransaction) {
            transactionOrSpan.finish();
          }

          return resultPromise;
        }
      };
      return intermediaryFunction();
    };
  };
};

/**
 * Decorator that marks a parameter as something to be included in the transaction name
 * @example
 * For a function with this signature,
 * `(param1: number, @SentryParam param2: string)`
 * doing `Reflect.getOwnMetadata(sentryParamsMetadataKey, target, propertyKey)` will return `[1]`
 *
 * for
 * `(@SentryParam param1: number, @SentryParam param2: string)` -> `[1,0]`
 */
export function SentryParam(
  target: any,
  propertyKey: string | symbol,
  parameterIndex: number,
) {
  // get the existing sentry params from the reflection metadata
  const sentryParams: SentryTracedParamsIndexes =
    Reflect.getOwnMetadata(sentryParamsMetadataKey, target, propertyKey) || [];
  // add the new parameter index to the the existing list
  sentryParams.push(parameterIndex);
  // override the metadata with the new list
  Reflect.defineMetadata(
    sentryParamsMetadataKey,
    sentryParams,
    target,
    propertyKey,
  );
}


example/src/services/MainService.ts
import { SentryTraced, SentryParam } from 'sentry-traced';
import fetch from 'node-fetch';
import * as Sentry from '@sentry/node';

/**
 * Example class to test the decorators
 */
export class MainService {
  @SentryTraced()
  async myMethod() {
    console.log('called myMethod');
    await this.myMethodChild1();
    await this.myMethodChild2();
  }

  @SentryTraced()
  private async myMethodChild1() {
    console.log('-- called myMethodChild1');
    await this.myMethodChild1Child1();
    await this.myMethodChild1Child2();
  }
  @SentryTraced()
  private async myMethodChild2() {
    console.log('-- called myMethodChild2');
    await this.myMethodChild2Child1();
  }
  @SentryTraced()
  private async myMethodChild1Child1() {
    console.log('---- called myMethodChild1Child1');
  }
  @SentryTraced()
  private async myMethodChild1Child2() {
    console.log('---- called myMethodChild1Child2');
  }
  @SentryTraced()
  private async myMethodChild2Child1() {
    console.log('---- called myMethodChild2Child1');
  }

  @SentryTraced()
  async exampleDistributedTracing() {
    const span = Sentry.getCurrentHub().getScope()?.getSpan();
    const result = await fetch('http://localhost:3000/other-endpoint', {
      headers: {
        'sentry-trace': span!.toTraceparent(),
      },
    });
    return await result.text();
  }
}


example/src/index.ts
import * as Sentry from '@sentry/node';
import { Integrations } from '@sentry/tracing';
import express from 'express';
import '@sentry/tracing'; // don't forget to import the tracing package as well
import { registerSentryInstance, withTracing } from 'sentry-traced';
import { MainService } from './services/MainService';
import { waitFor } from './utils';
import { SecondService } from './services/SecondService';

const app = express();
const port = 3000;

/**
 * Initialize Sentry with dsn and integrations
 */
Sentry.init({
  dsn: 'https://c66a2c25e42a4914a78354af798c3174@o1045421.ingest.sentry.io/6719428',
  integrations: [new Integrations.Express()],
  tracesSampleRate: 1.0,
});

/**
 * Register Sentry instance to be used in decorators.
 * Related to 'sentry-traced' library
 */
registerSentryInstance(Sentry);

/**
 * Initialize main service and second service (dummy services to test decorators)
 */
const mainService = new MainService();
const secondService = new SecondService();

app.get('/', async (_, res) => {
  await mainService.myMethod();
  await mainService.exampleDistributedTracing();
  res.send('Executed functions. Check Sentry dashboard!');
});

app.get('/other-endpoint', async (req, res) => {
  await waitFor(100);
  // get the sentry trace from the request header
  const sentryTrace = req.headers['sentry-trace'] as string;
  console.log('other-endpoint, received sentry trace header', { sentryTrace });
  // call the method with tracing
  const result = withTracing(sentryTrace, {
    op: 'otherEndpoint',
    description: 'transaction from other endpoint',
  })(secondService.myMethod)(123, 'param2', 'param3');

  res.send(result);
});

app.listen(port, () => {
  console.log(`Example app listening  http://localhost:${port}`);
});
